Copy SSL_CERT_FILE to studio internal and persist env var

smacfarlane opened the pull request:
Closes #6900 

This will copy an ssl certificate specified by the SSL_CERT_FILE env variable to the inside of the studio, and set the variable on the inside of the studio to the  internal cache.

@chefsalim  I think we'll want to pull the certificate copy back into `studio::enter` in the Rust code.  It will cut down on the duplication between shell/powershell and I _think_ it will need to live there anyways in order for the Docker studio to have the same functionality. 

I'm going to work on the powershell implementation and add some tests in the mean time. 

Signed-off-by: Scott Macfarlane <smacfarlane@chef.io>

mwrock commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +44,29 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+        debug!("Caching SSL_CERT_FILE {}", &cert_file);
+
+        if let Err(err) = stdfs::copy(cert_file, &cache_file) {
+            debug!("Unable to cache SSL_CERT_FILE: {}", err);
```
seems like this should be a warn or even error

mwrock commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +44,29 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+            debug!("Unable to cache SSL_CERT_FILE: {}", err);
+        }
+    } else {
+        debug!("Invalid path for SSL_CERT_FILE: {}", &cert_file);
```
I think this should be a warn or error too and I'd also add something like "check to make sure that file exists"

smacfarlane replied:
@mwrock @baumanj Do you think it is worth passing the `UI` into this function so we can emit this information as part of the normal output, or is `warn!` sufficient?

mwrock replied:
You could pass the `UI` which I think is better than `warn!` but perhaps even better whould be to have `cache_ssl_cert_file` return a `Result` and then `start` deals with the err using its `UI`.

chefsalim replied:
Note there is a separate card to plumb in the cert status into the UI so might be better to do that work together instead of in this PR

mwrock commented on components/studio/bin/hab-studio.ps1:
```diff
@@ -455,6 +455,17 @@ function Set-Secrets {
+  if($env:SSL_CERT_FILE) {
+    try {
+      $cert_filename = (Get-Item $env:SSL_CERT_FILE).Name
+      $env:SSL_CERT_FILE = Join-Path "Habitat:\hab\cache\ssl" $cert_filename
```
similar to linux it wil be scoped for the current process and all child processes. When you enter a studio it starts a new pwsh process so this will not leak into the parent (non-studio) environment.

mwrock commented on components/studio/bin/hab-studio.ps1:
```diff
@@ -455,6 +455,17 @@ function Set-Secrets {
+  if($env:SSL_CERT_FILE) {
+    try {
+      $cert_filename = (Get-Item $env:SSL_CERT_FILE).Name
+      $env:SSL_CERT_FILE = Join-Path "Habitat:\hab\cache\ssl" $cert_filename
```
Should be `$env:SSL_CERT_FILE = Join-Path $env:HAB_CACHE_SSL_PATH $cert_filename`
Only powershell understands the `Habitat:\` drive but `$env:HAB_CACHE_SSL_PATH` is the cannonical path.

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +44,29 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+                return;
+            }
+        };
+        let cache_file = cert_cache_dir.join(&cert_filename);
```
Since we only execute this code if `cert_path.is_file()` returns true, `cert_path.file_name()` is guaranteed to return `Some(_)`, meaning we can eliminate the whole `match` block and just do:
```rust
        let cache_file = cert_cache_dir.join(cert_path.file_name().unwrap());
```
or, to be a bit more careful:
```rust
        let cert_filename = cert_path.file_name().expect("is file");
        let cache_file = cert_cache_dir.join(cert_filename);
```

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +44,29 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+            }
+        };
+        let cache_file = cert_cache_dir.join(&cert_filename);
+        debug!("Caching SSL_CERT_FILE {}", &cert_file);
```
```suggestion
        debug!("Caching SSL_CERT_FILE {}", cert_file);
```

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +44,29 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+            debug!("Unable to cache SSL_CERT_FILE: {}", err);
+        }
+    } else {
+        debug!("Invalid path for SSL_CERT_FILE: {}", &cert_file);
```
```suggestion
        debug!("Invalid path for SSL_CERT_FILE: {}", cert_file);
```

chefsalim commented on components/studio/bin/hab-studio.sh:
```diff
@@ -882,6 +881,12 @@ chroot_env() {
 
+  if [ -n "${SSL_CERT_FILE:-}" ] \
+  && [ -z "${NO_MOUNT}" ] \
+  && [ -z "${NO_ARTIFACT_PATH}" ]; then 
```
Hmm, the NO_ARTIFACT_PATH test seems unnecessary.  I'd be concerned about changing an existing var that might break backward compat.  Maybe just remove the NO_ARTIFACT_PATH line here?

baumanj replied:
I can't seem to get `-N` or setting the `NO_ARTIFACT_PATH` environment variable to actually do what it purports to do:
```
➤ hab studio --help
…
COMMON FLAGS:
…
    -N  Do not mount the source artifact cache path into the Studio (default: mount the path)
ENVIRONMENT VARIABLES:
…
    NO_ARTIFACT_PATH       If set, do not mount the source artifact cache path (`-N' flag overrides)
```
```
➤ hab studio rm
…
➤ env DEBUG=1 hab studio -N enter
…
[1][default:/src:0]# ls /hab/cache/artifacts/
core-acl-2.2.53-20190115012136-x86_64-linux.hart             core-gzip-1.9-20190115013612-x86_64-linux.hart               core-openssl-1.0.2r-20190305210149-x86_64-linux.hart
…
```
even though when I grep the debug output, I see:
```
+ NO_ARTIFACT_PATH=true
```

So, we should probably figure out whether that's a bug and if it has impact here. If it's only supposed to apply to `/hab/cache/artifacts` and will leave `/hab/cache/ssl` alone, I agree that we should just omit that check here, but I got the sense that it may have been intended to control the mounting of the entire `/hab/cache` directory and that the name was from when we didn't have other things cached, but I'm really not sure.

smacfarlane replied:
This looks like a copy/paste gone wrong,  it should be `NO_CERT_PATH`, which I do think we want in this series of checks.  

@baumanj  I wasn't able to replicate the behavior you were seeing of `NO_ARTIFACT_PATH`, would you mind opening an issue on that?

baumanj replied:
What do you see when you pass `-N` or set `NO_ARTIFACT_PATH`, @smacfarlane ?

smacfarlane replied:
@baumanj If I enter an existing studio, I see an empty directory. If I destroy and create as you did, then I see the TDEPS of the studio, launcher and supervisor which is what I would expect.   `NO_ARTIFACT_CACHE` only controls the mounting of the cache from the host.  hab will still use `/hab/cache/artifacts` to cache everything it downloads, but the artifacts aren't shared with the host so will need to be fetched from builder and won't saved after the studio is removed.   Sorry I wasn't more clear this morning.

smacfarlane commented:
> The powershell looks good to me. Does `hab` spit out the `warn!` with the default logging level?

It does not,  but I'd like to address that in #6799

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -272,3 +295,89 @@ mod inner {
+    }
+
+    #[test]
+    fn cache_ssl_cert_file_invalid_cert_cache_dir() -> std::io::Result<()> {
```
Sorry, I'm not sure what I was thinking with suggesting this particular test. Since `cache_ssl_cert_file` doesn't report any errors, there's nothing observable that could happen with passing an invalid `cert_cache_dir`. We can probably remove this test altogether.

baumanj commented on components/hab/src/command/studio/docker.rs:
```diff
@@ -339,4 +364,30 @@ mod tests {
+                                                          .to_str()
+                                                          .unwrap()
+                                                          .to_string();
+        assert_eq!(std::env::var(SSL_CERT_FILE_ENVVAR), Ok(internal_cert_path));
```
This test would be more valuable if `mnt_prefix` weren't an empty string. As is it is now, we can break `update_ssl_cert_file_envvar` by changing
```rust
env::set_var(SSL_CERT_FILE_ENVVAR,
             Path::new(mnt_prefix).join(CACHE_SSL_PATH)
                                  .join(cert_file_name));
```
to
```rust
env::set_var(SSL_CERT_FILE_ENVVAR,
             Path::new(CACHE_SSL_PATH).join(cert_file_name));
```
and the test still passes

smacfarlane commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -272,3 +295,89 @@ mod inner {
+
+        Ok(())
+    }
+}
```
I've added a guard in the `cache_ssl_cert_file` function, but I'm not sure how we could test that.  Since `fs::copy` isn't guaranteed to truncate the file, inspecting the contents would likely intermittently fail.  There aren't any other side effects that I can think of to inspect, since the function is essentially a no-op in this case.

baumanj replied:
A test that intermittently fails in the face of incorrect logic is preferable to no test at all. Even better would be making `cache_ssl_cert_file` return a `Result` so we can test that we positively catch the conditions where the copy would refer to the same file (whether identical paths, symlinks or hard links).

smacfarlane commented:
@baumanj  I think I've addressed all your feedback, with the exception of the same-file test.  Any advice on how to proceed on that one would be appreciated.

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +45,22 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
 }
 
+fn cache_ssl_cert_file(cert_file: &str, cert_cache_dir: &Path) {
+    let cert_path = Path::new(&cert_file);
```
I've found myself getting confused when reading the logic of this function, and I think a few naming changes may make it easier to follow:
1. `cert_file` → `file_to_cache` or `src_cert`, etc.
1. `cert_cache_dir` → `dest_dir`
1. When converting `cert_file` of type `&str` → `Path`, re-bind the same name rather than introduce `cert_path`* (this is allowed and idiomatic) or make the function take a `Path` as input rather than `&str`.

These are just suggestions; not required changes

\* Since rust has such strong typing, in general I find it useful to put only the purpose of the data in the variable name and leave out words that describe the type.

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -88,6 +110,14 @@ pub fn start(ui: &mut UI, args: &[OsString]) -> Result<()> {
+        match cache_ssl_cert_file(&ssl_cert_file, &ssl_path) {
+            Ok(_) => {}
+            Err(err) => warn!("Unable to cache SSL_CERT_FILE: {}", err),
+        }
```
```rust
        if let Err(err) = cache_ssl_cert_file(&ssl_cert_file, &ssl_path) {
            warn!("Unable to cache SSL_CERT_FILE: {}", err);
        }
```
is a tad simpler/more idiomatic; but the match approach works too

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -1,23 +1,27 @@
             ORIGIN_ENVVAR};
 
 use habitat_core::AUTH_TOKEN_ENVVAR;
+use same_file::is_same_file;
```
I'm getting
> error[E0432]: unresolved import `same_file`

when I try this. Is there a change to the cargo.toml that didn't get pushed up, by any chance?

smacfarlane replied:
:man_facepalming: Fixed!

baumanj commented on components/hab/src/command/studio/enter.rs:
```diff
@@ -43,6 +47,24 @@ fn set_env_var_from_config(env_var: &str, config_val: Option<String>, sensitive:
+                                                  .to_string()));
+    }
+
+    debug!("Caching SSL_CERT_FILE {}", cert_file);
```
In case of errors, this would probably be more useful at the top of the function and also inclusive of the destination. Something like
```suggestion
    debug!("Caching SSL_CERT_FILE {:?} -> {:?}",
           cert_file, cert_cache_dir);
```
